gax/src/main/java/com/google/api/gax/rpc/mtls/MtlsProvider.java

```diff
@@ -0,0 +1,191 @@
+import com.google.api.client.json.gson.GsonFactory;
+import com.google.api.client.util.SecurityUtils;
+import com.google.api.core.BetaApi;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.security.GeneralSecurityException;
+import java.security.KeyStore;
+import java.util.List;
+
+/**
+ * Provider class for mutual TLS. It is used to configure the mutual TLS in the transport with the
+ * default client certificate on device.
+ */
+@BetaApi
+public class MtlsProvider {
+  interface EnvironmentProvider {
+    String getenv(String name);
+  }
+
+  static class SystemEnvironmentProvider implements EnvironmentProvider {
+    @Override
+    public String getenv(String name) {
+      return System.getenv(name);
+    }
+  }
+
+  interface ProcessProvider {
+    public Process createProcess(InputStream metadata) throws IOException;
+  }
+
+  static class DefaultProcessProvider implements ProcessProvider {
+    @Override
+    public Process createProcess(InputStream metadata) throws IOException {
+      if (metadata == null) {
+        return null;
+      }
+      List<String> command = extractCertificateProviderCommand(metadata);
+      return new ProcessBuilder(command).start();
+    }
+  }
+
+  private static final String DEFAULT_CONTEXT_AWARE_METADATA_PATH =
+      System.getProperty("user.home") + "/.secureConnect/context_aware_metadata.json";
+
+  private String metadataPath;
+  private EnvironmentProvider envProvider;
+  private ProcessProvider processProvider;
+
+  public enum UseMtlsEndpoint {
+    NEVER,
+    AUTO,
+    ALWAYS;
+  }
+
+  @VisibleForTesting
+  MtlsProvider(
+      EnvironmentProvider envProvider, ProcessProvider processProvider, String metadataPath) {
+    this.envProvider = envProvider;
+    this.processProvider = processProvider;
+    this.metadataPath = metadataPath;
+  }
+
+  public MtlsProvider() {
+    this(
+        new SystemEnvironmentProvider(),
+        new DefaultProcessProvider(),
+        DEFAULT_CONTEXT_AWARE_METADATA_PATH);
+  }
+
+  /**
+   * Returns if mutual TLS client certificate should be used. If the value is true, the key store
+   * from {@link #getKeyStore()} will be used to configure mutual TLS transport.
+   */
+  public boolean useMtlsClientCertificate() {
+    String useClientCertificate = envProvider.getenv("GOOGLE_API_USE_CLIENT_CERTIFICATE");
+    return "true".equals(useClientCertificate);
+  }
+
+  /**
+   * Returns AUTO, NEVER or ALWAYS. NEVER means always use regular endpoint; ALWAYS means always use
+   * mTLS endpoint; AUTO means auto switch to mTLS endpoint if client certificate exists and should
+   * be used.
+   */
+  public UseMtlsEndpoint useMtlsEndpoint() {
+    String useMtlsEndpoint = envProvider.getenv("GOOGLE_API_USE_MTLS_ENDPOINT");
+    if ("never".equals(useMtlsEndpoint)) {
```

Comment:
Consider something like:
```
private static final MTLS_ENDPOINT_POLICY_NAMES = 
    Stream.of(UseMtlsEndpoint.values()).map(UseMtlsEndpoint::name).collect(Collectors.toSet());

public UseMtlsEndpoint useMtlsEndpoint() {
 String useMtlsEndpoint = envProvider.getenv("GOOGLE_API_USE_MTLS_ENDPOINT").toUpperCase();
 if (MTLS_ENDPOINT_POLICY_NAMES.contains(useMtlsEndpoint)) {
   return UseMtlsEndpoint.valueOf(useMtlsEndpoint);
 } 
 return UseMtlsEndpoint.AUTO;
}

Reply:
Java 7 doesn't support stream, so I am sure if there is a clean way to do it this way in Java 7. I will leave it the way it is for now.